### ISSUE TITLE
Request for adding ligatures for template engines

### ISSUE THREAD

**Issue by hakatashi** [tonsky/FiraCode#75]
This is awesome coding font ever!

I suggest that it would be nice to have ligatures for some template engines, such as [EJS](http://www.embeddedjs.com/), [ERB](http://apidock.com/ruby/ERB), [ECT](http://ectjs.com/), [Jinja](http://jinja.pocoo.org/), [Twig](http://twig.sensiolabs.org/), [Smarty](http://www.smarty.net/), and native PHP. It is something like...

* `<%`, `<%=`, `<%#`, `%>`: Used in EJS, ERB, and ECT
* `<%-`: Used in ECT
* `{%`, `%}`, `{#`, `#}`, `{{`, `}}`: Used in Jinja and Twig
* `{*`, `*}`: Used in Smarty
* `<?`, `?>`: Used in PHP

**Comment by flying-sheep** [comment]
jinja also has a [semi-official set of delimiters for LaTeX](http://flask.pocoo.org/snippets/55/):

* `((*` `*))`
* `((=` `=))`
* `(((` `)))`

**Comment by brunobg** [comment]
[Mustache](http://mustache.github.com/) also would be nice: `{{`, `}}`, `{{{`, `}}}`, `{{<`

**Comment by MadcapJake** [comment]
Perl 6 uses double braces (`{{`) as well for long-form strings.

Perl 6 and Polymer.js use the double square brackets (`[[`) as well (strings and one-way bindings, respectively).

**Comment by tonsky** [comment]
Any idea how any of these should look like?

**Comment by MadcapJake** [comment]
Brackets/braces could just be squished like exclamation and question marks.

**Comment by flying-sheep** [comment]
apert from the obvious central line connecting the curlies, we could also try connecting them on top and bottom, like:

![brackets](https://cloud.githubusercontent.com/assets/291575/11951691/43e973c8-a893-11e5-814f-f8e587c9f1ea.png)

**Comment by brunobg** [comment]
flying-sheep's idea is nice. Possible idea: making the outer braces/brackets slightly smaller vertically than the inner ones, which would make it even clearer.

**Comment by coreh** [comment]
I like that, it looks like ⟦ and ⟧.

**Comment by weavejester** [comment]
The only potential problem I can see with this is that the ligatures might look odd on data structures like `[a [b c]]` or `{a {b c}}`. The ending `]]` and `}}` on those examples might make things look a little uneven if turned into a ligature.

**Comment by flying-sheep** [comment]
sure, but this already happens.

only solution is to modify your editor’s syntax highlighter to render it appropriately. e.g. in atom, you could tokenize

```rust
Vec<Hurr<Durr>>;
y << z;
```

(using pseudo-LaTeX instead of HTML to make it clear)

either greedily, which preserves all ligatures

```latex
Vec\op{<}Hurr\op{<}Durr\op{>>};
y \op{<<} z;
```

or lazily, which preserves none

```latex
Vec\op{<}Hurr\op{<}Durr\op{>}\op{>};
y \op{<}\op{<} z;
```

or smartly, which requires more work:

```latex
Vec\op{<}Hurr\op{<}Durr\op{>}\op{>};
y \op{<<} z;
```

**Comment by weavejester** [comment]
That's true, and I'd forgotten about the use of `>>` to denote some type signatures.

I'm not sure whether Emacs supports breaking up ligatures in the same way; however, I've just tried `Vec<Hurr<Durr>>` in my editor, and the ending `>>` is subtle enough that it doesn't appear uneven at all. If the `]]` and `}}` etc. ligatures are similarly subtle, then perhaps it doesn't matter.

**Comment by flying-sheep** [comment]
well, the only way i can imagine that a TUI emacs could do that is by controlling the sequence of color changes: <kbd>\x1b[31m</kbd><kbd>&gt;</kbd><kbd>&gt;</kbd><kbd>\x1b[39;49m</kbd> or <kbd>\x1b[31m</kbd><kbd>&gt;</kbd><kbd>\x1b[39;49m</kbd><kbd>\x1b[31m</kbd><kbd>&gt;</kbd><kbd>\x1b[39;49m</kbd> (i.e. by resetting the fg color between the symbols or not).

and then the terminal emulator needs to interpret this in the same way and only render ligatures if they are’t interrupted by a color change command

**Comment by igorsantos07** [comment]
I would add that, if there's going to be ligatures for `<?` and `<%=`, there must be for `<?=` as well. The issue I've just pointed out lists other very used PHP tokens.

**Comment by madcampos** [comment]
I'd really would like to see this.
for the design idea things like `<%` could follow the same pattern as `<$` or `</` that joins both characters.
and for `[[` the inner one could be smaller witch also works for "inner structures" like an array of arrays as: `[[1,2], [3,4]]`

**Comment by flying-sheep** [comment]
rather the outer one would be enlarged, else it would look unbalanced and confusing

**Comment by Blaimi** [comment]
some ideas:
![firacode_75](https://cloud.githubusercontent.com/assets/1222379/16849519/13b98a9e-49fb-11e6-96b3-e039e54dd146.png)

`<%=`, `<%#` and `<%-` depend on the ligatur for `<%` on their left.

I’d prefer
* `<?`: A
* `<?=` and `<?=$`: B
* `?>`: A
* `<%` and `%>`: B
* `{*`: both are ugly. A bigger `*` should be enough like already in `<*`

**Comment by igorsantos07** [comment]
Agree with almost all choices by @Blaimi, although for `?>` B sounds more natural - leave the dot where it is, instead of mangling it in the next symbol :)
I'm no designer here nor I understand of fonts, plus, those are sketches... so take my point with a grain of salt haha

Finally: maybe `{*` could be done with a bigger asterisk, joining both the top and bottom with the begin and end of the braces line?

**Comment by RajcsanyiZ** [comment]
![image](https://user-images.githubusercontent.com/1152691/97747319-11e8b700-1aec-11eb-94f1-7a7966c5ff68.png)

**Comment by RajcsanyiZ** [comment]
![image](https://user-images.githubusercontent.com/1152691/97747388-2f1d8580-1aec-11eb-923d-85afdb18ea07.png)

**Comment by RajcsanyiZ** [comment]
![image](https://user-images.githubusercontent.com/1152691/97752790-0352cd80-1af5-11eb-89de-ac905d692343.png)

**Comment by RajcsanyiZ** [comment]
I think the php close tag is not important, but the opening tag is very cool this way:

![image](https://user-images.githubusercontent.com/1152691/97812297-b439a380-1c80-11eb-8e6d-941052630e1a.png)

**Comment by RajcsanyiZ** [comment]
I worked out a little nicer:
![image](https://user-images.githubusercontent.com/1152691/97812710-70946900-1c83-11eb-8f8a-3a7fc0c65e15.png)

**Comment by RajcsanyiZ** [comment]
A little strong, but maybe thought-provoking.

![image](https://user-images.githubusercontent.com/1152691/97813218-ecdc7b80-1c86-11eb-8c4f-b4e450e9f8f0.png)

**Comment by RajcsanyiZ** [comment]
> some ideas:
> ![firacode_75](https://cloud.githubusercontent.com/assets/1222379/16849519/13b98a9e-49fb-11e6-96b3-e039e54dd146.png)
> 
> `<%=`, `<%#` and `<%-` depend on the ligatur for `<%` on their left.
> 
> I’d prefer
> 
> * `<?`: A
> * `<?=` and `<?=$`: B
> * `?>`: A
> * `<%` and `%>`: B
> * `{*`: both are ugly. A bigger `*` should be enough like already in `<*`

![image](https://user-images.githubusercontent.com/1152691/97994520-acdfda80-1de5-11eb-8299-ac6536d4fe1f.png)

**Comment by RajcsanyiZ** [comment]
Maybe it is a cleaner format:
![image](https://user-images.githubusercontent.com/1152691/97996030-95a1ec80-1de7-11eb-95fc-43d77430be00.png)
